src/orange/OrangeParams.cc

```diff
@@ -95,6 +95,65 @@ OrangeInput input_from_json(std::string filename)
     return result;
 }
 
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the maximum num of universe levels within the geometry
+ *
+ * A single-universe geometry will have a max_num_levels of 1.
+ * This is a recursive function, so for external callers, uid should be
+ * the root universe id.
+ */
+size_type calc_max_num_levels(HostVal<OrangeParamsData> const& data,
+                              UniverseId const& uid)
```

Comment:
I think this calculation would be easier to do on the input data, wouldn't it? Or is your thought that this can be reused even if we're building the shapes locally? Because it's using a lot of different low-level data structures in `OrangeParams`, it's going to have a relatively high maintenance burden compared to the simpler input abstractions.

Reply:
I figured it would better to operate on the final, already-built geometry since there may be different avenues for which the geometry is constructed. I see what you are saying about the data structures, but it doesn't seem to me that these would be any more likely to change than the ones on the input side.